zk/stages/stage_sequence_execute.go

```diff
@@ -266,36 +266,44 @@ func SpawnSequencingStage(
 					}
 
 					if anyOverflow {
+						// remove the last attempted counters as we may want to continue processing this batch with other transactions
+						batchCounters.RemovePreviousTransactionCounters()
+
 						if batchState.isLimboRecovery() {
 							panic("limbo transaction has already been executed once so they must not overflow counters while re-executing")
 						}
 
 						if !batchState.isL1Recovery() {
-							log.Info(fmt.Sprintf("[%s] overflowed adding transaction to batch", logPrefix), "batch", batchState.batchNumber, "tx-hash", txHash, "has-any-transactions-in-this-batch", batchState.hasAnyTransactionsInThisBatch)
 							/*
 								There are two cases when overflow could occur.
-								1. The block DOES not contains any transactions.
+								1. The block DOES not contain any transactions.
 									In this case it means that a single tx overflow entire zk-counters.
 									In this case we mark it so. Once marked it will be discarded from the tx-pool async (once the tx-pool process the creation of a new batch)
+									Block production then continues as normal looking for more suitable transactions
 									NB: The tx SHOULD not be removed from yielded set, because if removed, it will be picked again on next block. That's why there is i++. It ensures that removing from yielded will start after the problematic tx
 								2. The block contains transactions.
-									In this case, we just have to remove the transaction that overflowed the zk-counters and all transactions after it, from the yielded set.
-									This removal will ensure that these transaction could be added in the next block(s)
+									In this case we make note that we have had a transaction that overflowed and continue attempting to process transactions
+									Once we reach the cap for these attempts we will stop producing blocks and consider the batch done
 							*/
 							if !batchState.hasAnyTransactionsInThisBatch {
+								// mark the transaction to be removed from the pool
 								cfg.txPool.MarkForDiscardFromPendingBest(txHash)
-								log.Trace(fmt.Sprintf("single transaction %s overflow counters", txHash))
+								log.Info(fmt.Sprintf("[%s] single transaction %s cannot fit into batch", logPrefix, txHash))
+							} else {
+								batchState.newOverflowTransaction()
+								log.Info(fmt.Sprintf("[%s] transaction %s overflow counters", logPrefix, txHash), "count", batchState.overflowTransactions)
+								if batchState.reachedOverflowTransactionLimit() {
+									log.Info(fmt.Sprintf("[%s] closing batch due to counters", logPrefix), "count", batchState.overflowTransactions)
+									runLoopBlocks = false
+									break LOOP_TRANSACTIONS
+								}
 							}
-
-							runLoopBlocks = false
-							break LOOP_TRANSACTIONS
 						}
-
-					}
-
-					if err == nil {
-						blockDataSizeChecker = &backupDataSizeChecker
-						batchState.onAddedTransaction(transaction, receipt, execResult, effectiveGas)
+					} else {
```

Comment:
I think we don't need the "else", because if "else" exists then this code will never execute during overflow but we need this code to execute during overflow.

Reply:
We don't want this to run on an overflow because we haven't added the tx to the block.  Without this in place it causes a panic trying to read the receipt for the tx that wasn't added.

Reply:
In case of recovery, we could have added a tx to a block so we need this to be executed.

Reply:
OK, will add in something here for that

Reply:
Then again.... we still don't want to execute this in the case of an overflow even if we are in recovery?

Reply:
Any occurence of an overflow will mean the tx wasn't added to the block/batch so we don't want to include it in the batch state

Reply:
Ok so what will we do in case of overflow and L1Recovery? We cannot revert l1 so we can only build the same batch therefore we must add it to batch state.

Reply:
You're right, if we have data on L1 then it is the truth regardless of what virtual counters think.  I'll update the code 😄

Reply:
I think the latest commit here covers us off.  We basically note that we have skipped the tx as we're not in L1 recovery